src/main/java/org/zalando/intellij/swagger/file/FileDetector.java

```diff
@@ -3,42 +3,60 @@
 import org.jetbrains.yaml.YAMLLanguage;
 import org.zalando.intellij.swagger.traversal.path.PathFinder;
 
+import java.util.Optional;
+
 public class FileDetector {
 
     private static final String SWAGGER_KEY = "swagger";
+    private static final String SWAGGER_VERSION = "[\"'](2\\.0)[\"']"; //Must be "2.0", yaml does allow '2.0'
     private static final String OPEN_API_KEY = "openapi";
-
+    private static final String OPEN_API_VERSION = ".*(3(\\.\\d+)+)(?![\\d\\.]).*"; //openapi uses semantic versioning so 3.*.* would be allowed.
     public boolean isMainSwaggerJsonFile(final PsiFile psiFile) {
-        return hasJsonRootKey(psiFile, String.format("$.%s", SWAGGER_KEY));
+        return hasJsonRootKey(psiFile, String.format("$.%s", SWAGGER_KEY), SWAGGER_VERSION);
     }
 
     public boolean isMainSwaggerYamlFile(final PsiFile psiFile) {
-        return hasYamlRootKey(psiFile, String.format("$.%s", SWAGGER_KEY));
+        return hasYamlRootKey(psiFile, String.format("$.%s", SWAGGER_KEY), SWAGGER_VERSION);
     }
 
     public boolean isMainOpenApiJsonFile(final PsiFile psiFile) {
-        return hasJsonRootKey(psiFile, String.format("$.%s", OPEN_API_KEY));
+        return hasJsonRootKey(psiFile, String.format("$.%s", OPEN_API_KEY), OPEN_API_VERSION);
     }
 
     public boolean isMainOpenApiYamlFile(final PsiFile psiFile) {
-        return hasYamlRootKey(psiFile, String.format("$.%s", OPEN_API_KEY));
+        return hasYamlRootKey(psiFile, String.format("$.%s", OPEN_API_KEY), OPEN_API_VERSION);
     }
 
-    private boolean hasYamlRootKey(final PsiFile psiFile, final String lookupKey) {
+    private boolean hasYamlRootKey(final PsiFile psiFile, final String lookupKey, final String lookupVersion) {
         final Language language = psiFile.getLanguage();
-
-        return YAMLLanguage.INSTANCE.is(language) && new PathFinder().findByPathFrom(lookupKey, psiFile).isPresent();
+        if(!YAMLLanguage.INSTANCE.is(language)){
+            return false;
+        }
+        return new PathFinder().findByPathFrom(lookupKey, psiFile)
+                .filter(psiElement -> hasVersion(psiElement, lookupVersion)).isPresent();
     }
 
-    private boolean hasJsonRootKey(final PsiFile psiFile, final String lookupKey) {
+    private boolean hasJsonRootKey(final PsiFile psiFile, final String lookupKey, final String lookupVersion) {
         final Language language = psiFile.getLanguage();
+        if(!JsonLanguage.INSTANCE.is(language)){
+            return false;
+        }
+        return new PathFinder().findByPathFrom(lookupKey, psiFile)
+                .filter(psiElement -> hasVersion(psiElement, lookupVersion)).isPresent();
+    }
 
-        return JsonLanguage.INSTANCE.is(language) && new PathFinder().findByPathFrom(lookupKey, psiFile).isPresent();
+    private boolean hasVersion(final PsiElement psiElement, final String lookupVersion) {
+        return Optional.ofNullable(psiElement.getLastChild())
```

Comment:
This looks overly complex for detecting the text, this would do:

```java
        return Optional.ofNullable(psiElement.getLastChild())
                .map(PsiElement::getText)
                .filter(text -> text.matches(lookupVersion))
                .isPresent();
```

Reply:
Ok awesome. I've really just started using Streams and the first time I've used Optionals for anything. I'll update the logic to simplify it.